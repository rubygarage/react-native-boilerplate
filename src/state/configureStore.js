import { createStore, applyMiddleware } from 'redux';
import { createLogicMiddleware } from 'redux-logic';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore, persistReducer } from 'redux-persist';
import { Settings } from 'luxon';
import AsyncStorage from '@react-native-community/async-storage';

import buildHttpClient from 'lib/httpClient';
import { rootHttpClient } from 'lib/httpClient/buildHttpClient';
import mountInterceptors from 'lib/httpClient/mountInterceptors';
import conceptsOperations from './concepts/operationsRoot';
import flashOperations from './flash/operations';
import rootReducer from './store';

const operations = [...conceptsOperations, ...flashOperations];

const configureStore = () => {
  const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
    whitelist: ['data', 'application'],
    timeout: null,
  };
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  const operationsDependencies = {
    httpClient: buildHttpClient(),
    rootHttpClient,
  };

  const logicMiddleware = createLogicMiddleware(operations, operationsDependencies);
  const middlewares = applyMiddleware(logicMiddleware);
  const enhancer = composeWithDevTools(middlewares);

  const store = createStore(persistedReducer, enhancer);
  const persistor = persistStore(store);
  store.logicMiddleware = logicMiddleware;
  store.httpClient = operationsDependencies.httpClient;

  mountInterceptors(store);

  if (module.hot) {
    module.hot.accept(() => {
      // eslint-disable-next-line global-require
      store.replaceReducer(require('./store').default);
    });
  }

  Settings.defaultLocale = store.getState().intl.locale;

  return { store, persistor };
};

export default configureStore;
