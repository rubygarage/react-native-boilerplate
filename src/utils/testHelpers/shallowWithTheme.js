import React from 'react';
import { ThemeProvider } from 'styled-components/native';
// eslint-disable-next-line import/no-extraneous-dependencies
import { shallow } from 'enzyme';

import mockedTheme from './mockedTheme';

export default (tree) => shallow(
  <ThemeProvider theme={mockedTheme}>
    {tree}
  </ThemeProvider>,
);
