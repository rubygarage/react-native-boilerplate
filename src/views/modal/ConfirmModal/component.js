import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import ModalView from 'views/shared/ModalView';
import ModalHeader from 'views/shared/ModalView/ModalHeader';
import ModalBody from 'views/shared/ModalView/ModalBody';
import ModalFooter from 'views/shared/ModalView/ModalFooter';
import ModalText from 'views/shared/ModalView/ModalText';
import Button from 'views/shared/Button';

const ConfirmModalComponent = ({
  isVisible, onCloseModal, title, body, confirm,
  dismiss, onModalHide, onShouldConfirm, theme,
}) => (
  <ModalView isVisible={isVisible} onModalHide={onModalHide}>
    <ModalHeader isClosable onClose={onCloseModal}>
      <FormattedMessage {...title} />
    </ModalHeader>
    <ModalBody>
      {body.map((text) => (
        <ModalText key={text.id}>
          <FormattedMessage {...text} />
        </ModalText>
      ))}
    </ModalBody>

    <ModalFooter>
      <Button
        isSmallWidth
        isOutline
        isMarginRight
        onPress={onCloseModal}
      >
        <FormattedMessage {...dismiss} />
      </Button>
      <Button
        fillColor={theme.colors.red500}
        onPress={onShouldConfirm}
      >
        <FormattedMessage {...confirm} />
      </Button>
    </ModalFooter>
  </ModalView>
);

ConfirmModalComponent.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  onCloseModal: PropTypes.func.isRequired,
  onModalHide: PropTypes.func.isRequired,
  onShouldConfirm: PropTypes.func.isRequired,
  title: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  confirm: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  body: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
  })).isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
  dismiss: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

ConfirmModalComponent.defaultProps = {
  dismiss: { id: 'shared.cancel' },
};

export default ConfirmModalComponent;
