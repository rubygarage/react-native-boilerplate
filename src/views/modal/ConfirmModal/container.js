import React, { Component } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withTheme } from 'styled-components/native';
import PropTypes from 'prop-types';

import ConfirmModalComponent from './component';

class ConfirmModal extends Component {
  state = { isShouldConfirm: false }

  handleShouldConfirm = () => {
    const { onCloseModal } = this.props;
    this.setState({ isShouldConfirm: true });
    onCloseModal();
  }

  onModalHide = () => {
    const { onDestroyModal, dispatch, confirmAction } = this.props;
    const { isShouldConfirm } = this.state;
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  }

  render() {
    return (
      <ConfirmModalComponent
        {...this.props}
        onShouldConfirm={this.handleShouldConfirm}
        onModalHide={this.onModalHide}
      />
    );
  }
}

ConfirmModal.propTypes = {
  onCloseModal: PropTypes.func.isRequired,
  onDestroyModal: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  confirmAction: PropTypes.shape().isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});

const enhance = compose(
  connect(null, mapDispatchToProps),
  withTheme,
);

export { ConfirmModal as ConfirmModalContainer };
export default enhance(ConfirmModal);
