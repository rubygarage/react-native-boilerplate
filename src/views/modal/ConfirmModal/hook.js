import { useContext, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { useDispatch } from 'react-redux';

const useContainer = (props) => {
  const theme = useContext(ThemeContext);

  const { onCloseModal, onDestroyModal, confirmAction } = props;

  const dispatch = useDispatch();

  const [isShouldConfirm, setShouldConfirm] = useState(false);

  const handleShouldConfirm = () => {
    setShouldConfirm(true);
    onCloseModal();
  };

  const onModalHide = () => {
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  };

  return {
    theme,
    handleShouldConfirm,
    onModalHide,
    setShouldConfirm,
    isShouldConfirm,
  };
};

export default useContainer;
