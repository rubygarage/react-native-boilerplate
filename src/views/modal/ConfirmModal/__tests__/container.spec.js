import React from 'react';
import configureStore from 'redux-mock-store';

import shallowWithTheme from 'utils/testHelpers/shallowWithTheme';
import diveTo from 'utils/testHelpers/diveToEnzyme';
import ConfirmModal, { ConfirmModalContainer } from '../container';

describe('ConfirmModal container', () => {
  const dispatch = jest.fn();
  const store = configureStore()({});
  store.dispatch = dispatch;
  const onCloseModal = jest.fn();
  const onDestroyModal = jest.fn();
  const confirmAction = { type: 'Test action' };
  const defaultProps = {
    store,
    isVisible: true,
    onCloseModal,
    onDestroyModal,
    confirmAction,
    title: { id: 'test_title_id' },
    body: [{ id: 'test_body_1' }, { id: 'test_body_1' }],
    confirm: { id: 'test_confirm_id' },
    dismiss: { id: 'test_dismiss_id' },
  };

  const wrapper = shallowWithTheme(<ConfirmModal {...defaultProps} />);
  const container = diveTo(wrapper, ConfirmModalContainer);
  const instance = container.instance();

  beforeEach(() => {
    instance.setState({ isShouldLeaveConference: false });
  });

  it('renders correctly default', () => {
    expect(container).toMatchSnapshot();
  });

  it('changes "isShouldConfirm" state and calls "onCloseModal"', () => {
    expect(instance.state.isShouldConfirm).toBeFalsy();
    instance.handleShouldConfirm();
    expect(instance.state.isShouldConfirm).toBeTruthy();
    expect(onCloseModal).toHaveBeenCalled();
  });

  it('does not call "dispatch" when the "isShouldConfirm" state is false', () => {
    instance.setState({ isShouldConfirm: false });
    instance.onModalHide();
    expect(dispatch).not.toHaveBeenCalled();
    expect(onDestroyModal).toHaveBeenCalled();
  });

  it('does not call "dispatch" when the "isShouldConfirm" state is true', () => {
    instance.setState({ isShouldConfirm: true });
    instance.onModalHide();
    expect(dispatch).toHaveBeenCalledWith(confirmAction);
    expect(onDestroyModal).toHaveBeenCalled();
  });
});
