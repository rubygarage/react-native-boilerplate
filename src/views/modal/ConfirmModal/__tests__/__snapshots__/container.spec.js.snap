// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal container renders correctly default 1`] = `
<ConfirmModalComponent
  body={
    Array [
      Object {
        "id": "test_body_1",
      },
      Object {
        "id": "test_body_1",
      },
    ]
  }
  confirm={
    Object {
      "id": "test_confirm_id",
    }
  }
  confirmAction={
    Object {
      "type": "Test action",
    }
  }
  dismiss={
    Object {
      "id": "test_dismiss_id",
    }
  }
  dispatch={[MockFunction]}
  isVisible={true}
  onCloseModal={[MockFunction]}
  onDestroyModal={[MockFunction]}
  onModalHide={[Function]}
  onShouldConfirm={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  theme={
    Object {
      "colors": null,
      "fonts": null,
      "sizes": null,
      "type": "Mocked theme",
    }
  }
  title={
    Object {
      "id": "test_title_id",
    }
  }
/>
`;
